module2-database-technologies/README.md

```diff
@@ -1 +1,372 @@
+[declarative programming
+language](https://365datascience.com/tutorials/sql-tutorials/sql-declarative-language/)
+that is commonly used to query for data. Not all relational databases
+necessarily use SQL, but it is by far the most common, so in practice, SQL
+database and relational database are synomyous terms.
+
+The most frequently used relational databases are PostgreSQL and MySQL. There
+are many offerings for SQL databases, and each may have slightly different
+syntax for SQL. However, if you understand the general concept of relational
+databases, you will be able to easily use different databases. 
+
+### Understanding relational database structure
+As mentioned previously, data in relational databases are usually organized in a
+tabular format, similar to spreadsheets.
+
+* A database has many tables
+* A table has many rows (also known as records)
+* A row has many columns (also known as fields).
+
+For example, the previously mentioned `Customer` table may look something like
+the following. Note that this is not any particular syntax but simply a
+visualization of the data.
+
+**id**|**first\_name**|**last\_name**|**registered\_at**
+:-----:|:-----:|:-----:|:-----:
+1|Ammar|Sammour|2012-01-02
+2|Halit|Batur|2012-01-03
+3|Shrreya|Bhatachaarya|2012-01-05
+
+The `Order` table may look something like so:
+
+**id**|**product**|**delivered**|**customer\_id**
+:-----:|:-----:|:-----:|:-----:
+1|Keyboard|TRUE|1
+2|Mouse|FALSE|1
+3|Cookies|TRUE|1
+4|Rice|TRUE|2
+
+On the `Order` table, note that there is a field called `customer_id`. This is
+called a **foreign key** (or sometimes join key). This column creates a
+**relation** between the `Customer` and `Order` tables: it tells us to which
+customer the order belongs. This is the core of relational databases: expressing
+relations between entities.
+
+This table says that Ammar has ordered a keyboard, mouse, and cookies. He has
+three orders, because there are three rows with `customer_id = 1`. Halit has ordered one item: rice (`customer_id = 2`).
+
+### Describing relationships
+We usually use the terms one-to-many and many-to-many to describe
+relationships between tables. 
+
+For example, a customer may have multiple orders. This is a one-to-many relationship,
+since one customer has many orders. It should be noted that the foreign key goes
+on the side that is "many". In the example above, note that the `customer_id`
+key is located on the `Order` table.
+
+For a many-to-many relationship, consider students and courses at a university.
+A student can take many courses; a course also has many students. Many-to-many
+relationships are [expressed slightly
+differently](https://dzone.com/articles/how-to-handle-a-many-to-many-relationship-in-datab) than using a single foreign key.
+
+See this [StackOverflow
+question](https://stackoverflow.com/questions/4601703/difference-between-one-to-many-and-many-to-one-relationship)
+for more details.
```

Comment:
I feel like this section can be elaborated a bit further, explaining each of the 3 types of relationships (one-to-one, one-to-many, many-to-many) either with tabular examples like the above section, or visual examples like [here](https://afteracademy.com/blog/what-are-the-different-types-of-relationships-in-dbms)